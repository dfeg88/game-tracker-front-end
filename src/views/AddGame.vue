<template>
  <v-container>
    <v-col
        cols="12"
        lg="4"
        offset-lg="4"
    >
      <v-row class="text-center">
        <v-col class="mb-4">
          <h1 class="display-1 font-weight-bold mb-3">
            Add Game
          </h1>

          <v-form id="create" @submit.prevent="saveGame">
            <v-text-field
                v-model="gameRequest.name"
                label="Name"
                required
                data-test="name"
            ></v-text-field>
            <v-select
                v-model="gameRequest.platform"
                label="Platform"
                :items=platforms
                required
                data-test="platform"
            ></v-select>
            <v-text-field
                v-model="gameRequest.publisher"
                label="Publisher"
                required
                data-test="publisher"
            ></v-text-field>
            <v-select
                v-model="gameRequest.genre"
                label="Genre"
                :items=genres
                required
                data-test="genre"
            ></v-select>
            <br>
            <v-slider
                v-model="gameRequest.numberOfPlayers"
                label="Number of Players"
                thumb-color="blue"
                thumb-label="always"
                :thumb-size="18"
                min="1"
                max="32"
                data-test="numberOfPlayers"
            ></v-slider>

            <v-menu
                ref="releaseDateCalendar"
                v-model="releaseDateCalendar"
                :close-on-content-click="false"
                transition="scale-transition"
                offset-y
                max-width="290px"
                min-width="auto"
            >
              <template v-slot:activator="{ on, attrs }">
                <v-text-field
                    v-model="gameRequest.releaseDate"
                    label="Release Date"
                    persistent-hint
                    prepend-icon="mdi-calendar"
                    v-bind="attrs"
                    v-on="on"
                    data-test="release-date"
                ></v-text-field>
              </template>
              <v-date-picker
                  v-model="gameRequest.releaseDate"
                  no-title
                  @input="releaseDateCalendar = false"
              ></v-date-picker>
            </v-menu>
            <v-file-input
                v-model="boxArt"
                accept=".jpg,.png"
                label="Box Art"
                prepend-icon="mdi-camera"
                hint=".jpg or .png only"
                persistent-hint
                data-test="box-art"
            ></v-file-input>
            <br>
            <v-btn
                type="submit"
                id="submit"
            >
              Submit
            </v-btn>
          </v-form>
        </v-col>
      </v-row>
    </v-col>
  </v-container>
</template>

<script>
import {api} from "@/service/Api";

export default {
  name: 'AddGame',

  data: () => ({
    gameRequest: {
      name: '',
      platform: '',
      genre: '',
      releaseDate: new Date().toISOString().substr(0, 10),
      numberOfPlayers: 1,
      publisher: '',
    },
    boxArt: null,
    releaseDateCalendar: false,
    platforms: [
      'PS4',
      'PS5'
    ],
    genres: [
      'Action',
      'Adventure',
      'Fantasy',
      'First-Person Shooter',
      'Sport'
    ],

  }),
  methods: {
    saveGame() {
      let formData = new FormData();
      formData.append('boxArt', this.boxArt);
      formData.append('gameRequest', new Blob([JSON.stringify(this.gameRequest)], {
        type: "application/json"
      }))

      api()
          .post('v1/game', formData, {
            headers: {
              'Content-Type': 'multipart/form-date'
            }
          })
          .then(() => this.$router.push({
            name: 'games'
          }))
          .catch(err => console.error(err))
    }
  }
}
</script>
