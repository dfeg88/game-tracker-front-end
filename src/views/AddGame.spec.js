import {createLocalVue, mount} from "@vue/test-utils";
import AddGame from "@/views/AddGame";
import Vuetify from "vuetify";

describe('game creation', () => {
  let wrapper;
  let mockSave;
  let vuetify;

  jest.mock('axios', () => ({
    post: Promise.resolve()
  }))

  beforeEach(() => {
    vuetify = new Vuetify();
    mockSave = jest.spyOn(AddGame.methods, 'saveGame');

    wrapper = mount(AddGame, {
      localVue: createLocalVue(),
      vuetify,
    });

  })

  it('renders with expected HTML structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  })

  it('submits form', async () => {
    await wrapper.find('input[data-test=name').setValue('game name');
    await wrapper.find('input[data-test=platform').setValue('platform');
    await wrapper.find('input[data-test=publisher').setValue('publisher');
    await wrapper.find('input[data-test=genre').setValue('genre');
    await wrapper.find('input[data-test=numberOfPlayers').setValue('numberOfPlayers');
    await wrapper.find('input[data-test=release-date').setValue('releaseDate');

    await wrapper.find('.v-form').trigger('submit.prevent');

    expect(mockSave).toHaveBeenCalled()
  })
})