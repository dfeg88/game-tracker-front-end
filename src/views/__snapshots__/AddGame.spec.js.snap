// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game creation renders with expected HTML structure 1`] = `
<div
  class="container"
>
  <div
    class="col-lg-4 offset-lg-4 col-12"
  >
    <div
      class="row text-center"
    >
      <div
        class="mb-4 col"
      >
        <h1
          class="display-1 font-weight-bold mb-3"
        >
          
          Add Game
        
        </h1>
         
        <form
          class="v-form"
          id="create"
          novalidate="novalidate"
        >
          <div
            class="v-input theme--light v-text-field"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-3"
                    style="left: 0px; position: absolute;"
                  >
                    Name
                  </label>
                  <input
                    data-test="name"
                    id="input-3"
                    required="required"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-input theme--light v-text-field v-select"
          >
            <div
              class="v-input__control"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="list-6"
                class="v-input__slot"
                role="button"
              >
                <div
                  class="v-select__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-6"
                    style="left: 0px; position: absolute;"
                  >
                    Platform
                  </label>
                  <div
                    class="v-select__selections"
                  >
                    <input
                      aria-readonly="false"
                      autocomplete="off"
                      data-test="platform"
                      id="input-6"
                      readonly="readonly"
                      required="required"
                      type="text"
                    />
                  </div>
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--append"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-menu-down theme--light"
                      />
                    </div>
                  </div>
                  <input
                    type="hidden"
                  />
                </div>
                <div
                  class="v-menu"
                >
                  <!---->
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-input theme--light v-text-field"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-11"
                    style="left: 0px; position: absolute;"
                  >
                    Publisher
                  </label>
                  <input
                    data-test="publisher"
                    id="input-11"
                    required="required"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-input theme--light v-text-field v-select"
          >
            <div
              class="v-input__control"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="list-14"
                class="v-input__slot"
                role="button"
              >
                <div
                  class="v-select__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-14"
                    style="left: 0px; position: absolute;"
                  >
                    Genre
                  </label>
                  <div
                    class="v-select__selections"
                  >
                    <input
                      aria-readonly="false"
                      autocomplete="off"
                      data-test="genre"
                      id="input-14"
                      readonly="readonly"
                      required="required"
                      type="text"
                    />
                  </div>
                  <div
                    class="v-input__append-inner"
                  >
                    <div
                      class="v-input__icon v-input__icon--append"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-menu-down theme--light"
                      />
                    </div>
                  </div>
                  <input
                    type="hidden"
                  />
                </div>
                <div
                  class="v-menu"
                >
                  <!---->
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <br />
           
          <div
            class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input__slider"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-19"
                  style="left: 0px; position: relative;"
                >
                  Number of Players
                </label>
                <div
                  class="v-slider v-slider--horizontal theme--light"
                >
                  <input
                    data-test="numberOfPlayers"
                    disabled="disabled"
                    id="input-19"
                    readonly="readonly"
                    tabindex="-1"
                    value="1"
                  />
                  <div
                    class="v-slider__track-container"
                  >
                    <div
                      class="v-slider__track-background primary lighten-3"
                      style="right: 0px; width: calc(100%);"
                    />
                    <div
                      class="v-slider__track-fill primary"
                      style="left: 0px; width: 0%;"
                    />
                  </div>
                  <div
                    aria-label="Number of Players"
                    aria-orientation="horizontal"
                    aria-readonly="false"
                    aria-valuemax="32"
                    aria-valuemin="1"
                    aria-valuenow="1"
                    class="v-slider__thumb-container v-slider__thumb-container--show-label blue--text"
                    data-test="numberOfPlayers"
                    role="slider"
                    style="left: 0%;"
                    tabindex="0"
                  >
                    <div
                      class="v-slider__thumb blue"
                    />
                    <transition-stub
                      name="scale-transition"
                    >
                      <div
                        class="v-slider__thumb-label-container"
                      >
                        <div
                          class="v-slider__thumb-label blue"
                          style="height: 18px; width: 18px; transform: translateY(-20%) translateY(-12px) translateX(-50%) rotate(45deg);"
                        >
                          <div>
                            <span>
                              1
                            </span>
                          </div>
                        </div>
                      </div>
                    </transition-stub>
                  </div>
                </div>
              </div>
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
           
          <div
            class="v-menu"
          >
            <!---->
          </div>
          <div
            class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-calendar theme--light"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label v-label--active theme--light"
                    for="input-24"
                    style="left: 0px; position: absolute;"
                  >
                    Release Date
                  </label>
                  <input
                    aria-expanded="false"
                    aria-haspopup="true"
                    data-test="release-date"
                    id="input-24"
                    role="button"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
           
          <div
            class="v-input theme--light v-text-field v-file-input"
          >
            <div
              class="v-input__prepend-outer"
            >
              <div
                class="v-input__icon v-input__icon--prepend"
              >
                <button
                  aria-label="prepend icon"
                  class="v-icon notranslate v-icon--link mdi mdi-camera theme--light"
                  type="button"
                />
              </div>
            </div>
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-28"
                    style="left: 0px; position: absolute;"
                  >
                    Box Art
                  </label>
                  <div
                    class="v-file-input__text"
                  />
                  <input
                    accept=".jpg,.png"
                    data-test="box-art"
                    id="input-28"
                    type="file"
                  />
                </div>
                <div
                  class="v-input__append-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--clear"
                  >
                    <button
                      aria-label="clear icon"
                      class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"
                      disabled="disabled"
                      type="button"
                    />
                  </div>
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  >
                    <div
                      class="v-messages__message"
                    >
                      .jpg or .png only
                    </div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
           
          <br />
           
          <button
            class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
            id="submit"
            type="submit"
          >
            <span
              class="v-btn__content"
            >
              
            Submit
          
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
